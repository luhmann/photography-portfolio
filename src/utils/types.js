import { arrayOf, shape, string, number } from 'prop-types';

export const locationType = shape({
  pathname: string.isRequired,
});

export const galleryInfoType = shape({
  title: string.isRequired,
  path: string.isRequired,
});

export const albumInfoType = shape({
  albumTitle: string.isRequired,
  galleries: arrayOf(galleryInfoType).isRequired,
});

export const imageType = shape({
  contentDigest: string.isRequired,
  fluid: shape({
    tracedSVG: string.isRequired,
    aspectRatio: number.isRequired,
    src: string.isRequired,
    srcSet: string.isRequired,
    srcWebp: string.isRequired,
    srcSetWebp: string.isRequired,
    sizes: string.isRequired,
  }).isRequired,
});

export const createImageTypeTestProps = (index, custom = {}) => ({
  contentDigest: `digest-${index}`,
  fluid: {
    tracedSVG: `data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' width='400' height='267' viewBox='0 0 400 267' version='1'%3e%3cpath d='tracedsvg-${index}' fill='lightgray' fill-rule='evenodd'/%3e%3c/svg%3e`,
    aspectRatio: 1.5,
    src: `/static/digest-${index}/size/mock-${index}.jpg`,
    srcSet: `/static/digest-${index}/size/mock-${index}-625.jpg 625w,\n/static/digest-${index}/size/mock-${index}-1250.jpg 1250w,\n/static/digest-${index}/size/mock-${index}-2250.jpg 2250w`,
    srcWebp: `/static/digest-${index}/size/mock-${index}.webp`,
    srcSetWebp: `/static/digest-${index}/size/mock-${index}-625.webp 625w,\n/static/digest-${index}/size/mock-${index}-1250.webp 1250w,\n/static/digest-${index}/size/mock-${index}-2250.webp 2250w`,
    sizes: '(max-width: 2250px) 100vw, 2250px',
  },
  ...custom,
});
