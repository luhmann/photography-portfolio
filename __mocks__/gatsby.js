const React = require('react');
const gatsby = jest.requireActual('gatsby');

const navigate = jest.fn();

module.exports = {
  ...gatsby,
  graphql: jest.fn(),
  navigate,
  Link: jest.fn().mockImplementation(({ to, ...rest }) =>
    React.createElement('a', {
      ...rest,
      href: to,
    })
  ),
  StaticQuery: jest.fn(),
};
