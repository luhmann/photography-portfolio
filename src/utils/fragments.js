export const GalleryImagesFragment = graphql`
  fragment GalleryImagesFragment on FileConnection {
    edges {
      node {
        childImageSharp {
          internal {
            contentDigest
          }
          sizes(maxWidth: 2500, quality: 75) {
            ...GatsbyImageSharpSizes_withWebp_tracedSVG
          }
        }
      }
    }
  }
`;

export const allGalleriesYamlFragment = graphql`
  fragment allGalleriesYamlFragment on RootQueryType {
    allGalleriesYaml {
      group(field: album) {
        fieldValue
        edges {
          node {
            title
            path
          }
        }
      }
    }
  }
`;

export const singleGalleryYamlFragment = graphql`
  fragment singleGalleryYamlFragment on RootQueryType {
    galleriesYaml(folderName: { eq: $folderName }) {
      title
    }
  }
`;
