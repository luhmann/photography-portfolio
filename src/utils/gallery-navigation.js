import { inc, dec, ifElse } from 'rambda';

const nextStepper = (index, total = 0) =>
  ifElse(() => inc(index) < total, () => inc(index), () => 0);

const prevStepper = (index, total = 0) =>
  ifElse(() => dec(index) >= 0, () => dec(index), () => dec(total));

export { nextStepper, prevStepper };
