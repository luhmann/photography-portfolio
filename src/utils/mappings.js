import { path, pathOr, pipe, map } from 'rambda';

const flattenAllGalleriesGraphQLResponse = response =>
  response.map(({ fieldValue: albumTitle, edges: galleries }) => ({
    albumTitle,
    galleries: galleries.map(path('node')),
  }));

export const mapAllGalleriesGraphQLResponse = pipe(
  pathOr([], 'allGalleriesYaml.group'),
  flattenAllGalleriesGraphQLResponse
);

export const mapGalleryImagesGraphQLResponse = pipe(
  pathOr([], 'data.allFile.edges'),
  map(
    pipe(
      pathOr([], 'node.childImageSharp'),
      img => ({
        fluid: img.fluid,
        contentDigest: path('internal.contentDigest', img),
      })
    )
  )
);

export const mapSingleGalleryYamlGraphQLResponse = path(
  'data.galleriesYaml.title'
);
