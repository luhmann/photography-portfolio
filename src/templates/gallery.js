import React from 'react';
import PropTypes from 'prop-types';
import Helmet from 'react-helmet';
import Img from 'gatsby-image';
import { graphql } from 'gatsby';
import KeyHandler from 'react-key-handler';
import { dec, ifElse, inc } from 'rambda';
import { compose, mapProps, withStateHandlers } from 'recompose';
import styled, { createGlobalStyle } from 'styled-components';
import { themeGet } from 'styled-system';
import { ReactComponent as PrevIcon } from '../assets/prev.svg';
import { ReactComponent as NextIcon } from '../assets/next.svg';
import { ContentContainer, Layout } from '../components';
import {
  mapGalleryImagesGraphQLResponse,
  mapSingleGalleryYamlGraphQLResponse,
} from '../utils/mappings';
import { mediaScreen } from '../theme';

const GalleryContainer = styled(ContentContainer)`
  padding: ${themeGet('space.6')};

  ${mediaScreen.md`
    padding: ${themeGet('space.3')};
    padding-top: ${themeGet('space.6')};
    will-change: scroll-position;
  `};
`;

// NOTE: the introduction of global classes is unfortunate here,
// but unavoidable because of the API the `gatsby-image`-component exposes
// it allows to pass `className` which styles the inner of the two divs the component renders
// it allows to pass `outerClassName` which styles the outer image wrapper
// for use with styled-components it would be preferrable if there was `className` and `innerClassName`
// because with styled-components (and css) we only can easily stile "downwards" into the DOM tree
// to work around this we define a global classname and apply it conditionally
const outerWrapperHiddenClassName = 'outerWrapper-hidden';
const GatsbyImageOverwriteStyles = createGlobalStyle`
  .gatsby-image-outer-wrapper {
    width: 100%;
    height: 100%;
  }

  .${outerWrapperHiddenClassName} {
    display: none;
    ${mediaScreen.md`
      display: block;
    `};
  }
`;

const Image = styled(Img).attrs(props => ({
  outerWrapperClassName: props.visible ? null : outerWrapperHiddenClassName,
}))`
  align-items: center;
  display: ${props => (props.visible ? 'flex' : 'none')};
  height: 100%;

  ${mediaScreen.md`
    display: flex;
    height: auto;
    margin-bottom: ${themeGet('space.3')};
  `};
`;

const Prev = styled.a`
  align-items: center;
  cursor: w-resize;
  display: flex;
  height: 100%;
  padding: ${themeGet('space.3')};
  position: absolute;
  left: 0;
  top: 0;
  width: 49vw;
  z-index: ${themeGet('zIndex.low')};

  ${mediaScreen.md`
    display: none;
  `};
`;

const Next = styled(Prev)`
  cursor: e-resize;
  left: auto;
  right: 0;
  justify-content: flex-end;
`;

const StyledPrevIcon = styled(PrevIcon)`
  cursor: pointer;
  height: ${themeGet('space.5')};
  width: ${themeGet('space.5')};
`;

const StyledNextIcon = styled(NextIcon)`
  cursor: pointer;
  height: ${themeGet('space.5')};
  width: ${themeGet('space.5')};
`;

const Gallery = ({ images, title, imageIndex, next, prev, location }) => (
  <Layout location={location}>
    <GatsbyImageOverwriteStyles />
    <GalleryContainer>
      <Helmet>
        <title>{`${title} - JF Dietrich Photography`}</title>
      </Helmet>
      <KeyHandler keyValue="ArrowRight" onKeyHandle={next} />
      <KeyHandler keyValue="ArrowLeft" onKeyHandle={prev} />
      <Prev onClick={prev}>
        <StyledPrevIcon />
      </Prev>
      {images.map((image, index) => (
        <Image
          key={image.contentDigest}
          visible={imageIndex === index}
          fluid={image.fluid}
          imgStyle={{
            objectFit: 'contain',
          }}
        />
      ))}
      <Next onClick={next}>
        <StyledNextIcon />
      </Next>
    </GalleryContainer>
  </Layout>
);

Gallery.propTypes = {
  images: PropTypes.arrayOf(
    PropTypes.shape({
      fluid: PropTypes.object.isRequired,
      contentDigest: PropTypes.string.isRequired,
    })
  ).isRequired,
  title: PropTypes.string.isRequired,
  imageIndex: PropTypes.number.isRequired,
  next: PropTypes.func.isRequired,
  prev: PropTypes.func.isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired,
  }).isRequired,
};

Gallery.displayName = 'Gallery';

export default compose(
  mapProps(props => ({
    ...props,
    images: mapGalleryImagesGraphQLResponse(props),
    title: mapSingleGalleryYamlGraphQLResponse(props),
  })),
  withStateHandlers(
    ({ initialImageIndex = 0 }) => ({
      imageIndex: initialImageIndex,
    }),
    {
      next: ({ imageIndex }, { images }) =>
        ifElse(
          () => inc(imageIndex) < images.length,
          () => ({ imageIndex: inc(imageIndex) }),
          () => ({ imageIndex: 0 })
        ),
      prev: ({ imageIndex }, { images }) =>
        ifElse(
          () => dec(imageIndex) >= 0,
          () => ({ imageIndex: dec(imageIndex) }),
          () => ({ imageIndex: dec(images.length) })
        ),
      showIndex: () => ({ nextIndex }) => ({
        imageIndex: nextIndex,
      }),
    }
  )
)(Gallery);

export const pageQuery = graphql`
  query Gallery($folderName: String!) {
    ...singleGalleryYamlFragment
    allFile(
      filter: { relativeDirectory: { eq: $folderName } }
      sort: { fields: [relativePath] }
    ) {
      ...GalleryImagesFragment
    }
  }
`;
