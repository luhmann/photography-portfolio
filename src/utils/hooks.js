import { useState, useEffect, useRef } from 'react';
import { pipe, path, find, findIndex, equals, head, defaultTo } from 'rambda';
import { navigate } from 'gatsby';

import { nextStepper, prevStepper } from 'utils/gallery-navigation';

export function useInterval(callback, delay) {
  const savedCallback = useRef();

  // remember the latest callback.
  useEffect(() => {
    savedCallback.current = callback;
  });

  // set up the interval.
  useEffect(
    () => {
      function tick() {
        savedCallback.current();
      }
      if (delay !== null) {
        let id = setInterval(tick, delay);
        return () => clearInterval(id);
      }
    },
    [delay]
  );
}

const getImageId = path('contentDigest');
const getCurrentIndexById = id => images =>
  findIndex(
    pipe(
      getImageId,
      equals(id)
    )
  )(images);

export function useGallery({ images = [], initialId }) {
  const total = images.length;
  const verifiedInitialId = pipe(
    find(
      pipe(
        getImageId,
        equals(initialId)
      )
    ),
    defaultTo(head(images)),
    getImageId
  )(images);
  const [id, setId] = useState(verifiedInitialId);

  useEffect(
    () => {
      setId(verifiedInitialId);
    },
    [verifiedInitialId]
  );

  if (total === 0) {
    navigate('/404');
    return {};
  }

  const getNextIndex = nextStepper(getCurrentIndexById(id)(images), total);
  const getPrevIndex = prevStepper(getCurrentIndexById(id)(images), total);
  const nextId = getImageId(images[getNextIndex()]);
  const prevId = getImageId(images[getPrevIndex()]);

  const next = () => setId(nextId);
  const prev = () => setId(prevId);

  return { next, prev, currentId: id, nextId, prevId };
}
