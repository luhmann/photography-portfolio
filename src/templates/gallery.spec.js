import React from 'react';
import { render } from 'react-testing-library';
import generateProps from 'react-generate-props';
import { StaticQuery } from 'gatsby';

import { createImageTypeTestProps } from 'utils/types';

import { Gallery } from './gallery';

const createTestProps = custom => ({
  ...generateProps(Gallery, { optional: true }),
  images: [
    createImageTypeTestProps(1),
    createImageTypeTestProps(2),
    createImageTypeTestProps(3),
    createImageTypeTestProps(4),
    createImageTypeTestProps(5),
    createImageTypeTestProps(6),
    createImageTypeTestProps(7),
  ],
  ...custom,
});

const testSelectors = Object.freeze({
  // NOTE: `gatsby-image` does not pass on data-attributes
  altGalleryImage: /Gallery title/,
  testIdNextButton: 'gallery-next',
  testIdPrevButton: 'gallery-prev',
});

beforeEach(() => {
  StaticQuery.mockImplementation(({ render }) => render({}));
});

test('should render all images', async () => {
  const props = createTestProps();
  const { getAllByAltText } = render(<Gallery {...props} />);

  expect(getAllByAltText(testSelectors.altGalleryImage).length).toEqual(
    props.images.length
  );
});

// NOTE: because of styled components the difference is only visible in the classname
// this should be resolved once we have a better snapshot-serializer for styled components
test('should show the second image, when clicking next', async () => {
  const props = createTestProps();
  const updatedProps = createTestProps({
    initialId: 'digest-2',
  });
  const { asFragment, rerender } = render(<Gallery {...props} />);

  const firstRender = asFragment();

  // NOTE: not optimal way of testing this, but as we are relying on real navigation for this the props will come from the outside
  // it would be better to do this by wrapping it in the reach-router-provider but as that is hidden by gatsby it will be fiddly
  rerender(<Gallery {...updatedProps} />);

  expect(firstRender).toMatchDiffSnapshot(asFragment());
});

test('should show last image, when clicking prev', () => {
  const props = createTestProps();
  const updatedProps = createTestProps({
    initialId: 'digest-7',
  });
  const { rerender, asFragment } = render(<Gallery {...props} />);

  const firstRender = asFragment();

  rerender(<Gallery {...updatedProps} />);

  expect(firstRender).toMatchDiffSnapshot(asFragment());
});
