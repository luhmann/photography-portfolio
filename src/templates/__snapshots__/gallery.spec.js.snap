// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show last image, when clicking prev 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -26,18 +26,18 @@
        class=\\"sc-bwzfXH sc-gzVnrw eRdYLe\\"
      >
        <a
          class=\\"sc-dnqmqq kFbRKc\\"
          data-testid=\\"gallery-prev\\"
-         href=\\"pathname#digest-7\\"
+         href=\\"pathname#digest-6\\"
        >
          <svg
            class=\\"sc-gZMcBi cxRXTt\\"
          />
        </a>
        <div
-         class=\\"sc-htoDjs ivNpRq gatsby-image-wrapper\\"
+         class=\\"sc-htoDjs ezRVes gatsby-image-wrapper\\"
          style=\\"position: relative; overflow: hidden;\\"
        >
          <div
            style=\\"width: 100%; padding-bottom: 66.66666666666667%;\\"
          />
@@ -273,11 +273,11 @@
  /static/digest-6/size/mock-6-1250.jpg 1250w,
  /static/digest-6/size/mock-6-2250.jpg 2250w\\" sizes=\\"(max-width: 2250px) 100vw, 2250px\\" /&gt;&lt;img src=\\"/static/digest-6/size/mock-6.jpg\\" alt=\\"Gallery title - Image 6\\" style=\\"position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/&gt;&lt;/picture&gt;
          </noscript>
        </div>
        <div
-         class=\\"sc-htoDjs ezRVes gatsby-image-wrapper\\"
+         class=\\"sc-htoDjs ivNpRq gatsby-image-wrapper\\"
          style=\\"position: relative; overflow: hidden;\\"
        >
          <div
            style=\\"width: 100%; padding-bottom: 66.66666666666667%;\\"
          />
@@ -315,11 +315,11 @@
          </noscript>
        </div>
        <a
          class=\\"sc-dnqmqq sc-iwsKbI jwjenV\\"
          data-testid=\\"gallery-next\\"
-         href=\\"pathname#digest-2\\"
+         href=\\"pathname#digest-1\\"
        >
          <svg
            class=\\"sc-gqjmRU fOxXHI\\"
          />
        </a>"
`;

exports[`should show the second image, when clicking next 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -26,18 +26,18 @@
        class=\\"sc-bwzfXH sc-gzVnrw eRdYLe\\"
      >
        <a
          class=\\"sc-dnqmqq kFbRKc\\"
          data-testid=\\"gallery-prev\\"
-         href=\\"pathname#digest-7\\"
+         href=\\"pathname#digest-1\\"
        >
          <svg
            class=\\"sc-gZMcBi cxRXTt\\"
          />
        </a>
        <div
-         class=\\"sc-htoDjs ivNpRq gatsby-image-wrapper\\"
+         class=\\"sc-htoDjs ezRVes gatsby-image-wrapper\\"
          style=\\"position: relative; overflow: hidden;\\"
        >
          <div
            style=\\"width: 100%; padding-bottom: 66.66666666666667%;\\"
          />
@@ -73,11 +73,11 @@
  /static/digest-1/size/mock-1-1250.jpg 1250w,
  /static/digest-1/size/mock-1-2250.jpg 2250w\\" sizes=\\"(max-width: 2250px) 100vw, 2250px\\" /&gt;&lt;img src=\\"/static/digest-1/size/mock-1.jpg\\" alt=\\"Gallery title - Image 1\\" style=\\"position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/&gt;&lt;/picture&gt;
          </noscript>
        </div>
        <div
-         class=\\"sc-htoDjs ezRVes gatsby-image-wrapper\\"
+         class=\\"sc-htoDjs ivNpRq gatsby-image-wrapper\\"
          style=\\"position: relative; overflow: hidden;\\"
        >
          <div
            style=\\"width: 100%; padding-bottom: 66.66666666666667%;\\"
          />
@@ -315,11 +315,11 @@
          </noscript>
        </div>
        <a
          class=\\"sc-dnqmqq sc-iwsKbI jwjenV\\"
          data-testid=\\"gallery-next\\"
-         href=\\"pathname#digest-2\\"
+         href=\\"pathname#digest-3\\"
        >
          <svg
            class=\\"sc-gqjmRU fOxXHI\\"
          />
        </a>"
`;
